libraries/node-core-library/src/test/Async.test.ts

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { Async } from '../Async';
+
+describe('Async', () => {
+  describe('mapLimitAsync', () => {
+    it('returns the same result as built-in Promise.all', async () => {
+      const array: number[] = [1, 2, 3, 4, 5, 6, 7, 8];
+      const fn: (item: number) => Promise<string> = async (item) => `result ${item}`;
+
+      expect(await Async.mapLimitAsync(array, 1, fn)).toEqual(await Promise.all(array.map(fn)));
```

Comment:
Consider the more typical way these parameters might be inlined:
```ts
await Async.mapLimitAsync(
  [1, 2, 3, 4, 5, 6, 7, 8],
  1,
  async (item: number): Promise<string> => {
    return `result ${item}`;
  }
)
```
The `1` might get buried under a large array.  For someone unfamiliar with this API, it's not entirely obvious that `1` is the "limit".  

@iclanton  What if we changed this API to take an options interface, like this:
```ts
await Async.mapAsync({
  parallelism: 1,
  action: async (item: number): Promise<string> => {
    return `result ${item}`;
  },
  items: [1, 2, 3, 4, 5, 6, 7, 8]
})
```

This might be more readable.  It lets us drop "limit" from the function name, and later maybe we could add other throttling options like this:
```ts
await Async.mapAsync({
  maxCallsPerSecond: 10,
  action: async (item: number): Promise<string> => {
    return `result ${item}`;
  },
  items: [1, 2, 3, 4, 5, 6, 7, 8]
})
```

What do you think?

Reply:
The `forEach` variant would look like this:
```ts
      await Async.forEachAsync({
        parallelism: 3,
        action: async (item: number): Promise<void> => {
          running++;
          await Async.sleep(1);
          maxRunning = Math.max(maxRunning, running);
          running--;
        },
        items: [1, 2, 3, 4, 5, 6, 7, 8]
      });
```